### ISSUE TITLE
RT #98905: factor and primes are limited to 2^32

### ISSUE THREAD

**Issue by briandfoy** [briandfoy/PerlPowerTools#404]
From DANAJ in [#98905 on rt.cpan.org](https://rt.cpan.org/Ticket/Display.html?id=98905):

Both factor and primes check that the inputs are 10 or fewer digits and less than 2^32.  It is unclear to me whether this is:

  1) intentional behavior that should always be maintained for compatibility.

or

  2) sanity checking the range, and since 64-bit Perl didn't come until after these were written, 32-bit was hard coded.

These are not hard to extend, albeit they're not super speedy for larger numbers, but Ben Tilly wrote some pretty nice code so not nearly as bad as one might expect.  I've been playing with alternate primes implementations -- it can be easily sped up 2x for large bases, and I'm sure more than that if we accept lots of turmoil.

For extra credit, we could even use Math::BigInt, but doing that for all inputs while maintaining decent speed might give too much bloat (at least for my coding it would).

**Comment by briandfoy** [comment]
Hey @mknos, is this one something that you are interested in?

**Comment by pjacklam** [comment]
I played with the `factor` program, and it puzzled me why it became so slow then the input was large. I didn't look closely into the implementation, because I knew I had a `factor` program that was much faster.

First I modified the original program to allow for larger inputs. I changed the test into the following, which allows input values up to 2¹¹² on systems built with quadmath and 2⁶⁴-1 on systems without quadmath, but with for 64 bit integers.

```
   unless ($n =~ /^\+?\d+$/) {
        warn "$0: `$n:' is not a valid positive integer\n";
        return;
    }

    if ($n % 2 == 0 && ($n - 1) % 2 == 0) {
        warn "$0: `$n:' is too large\n";
        return;
    }
```
Then I improved my program and ran it against the current `factor` program. Below are the results. The execution times are in seconds. My program is about 4 times as large as the current `factor` program, which I assume explains why my program is slightly slower for the smallest input values. Note that I didn't test the original `factor` with the largest inputs. It would have taken many hours. I'll update this entry if I get the data.

|  n | largest n-digit prime | factor (new) | factor (orig) |
|  - | - | - | - |
|  1 | 7 | 0.11 | 0.03 |
|  2 | 97 | 0.11 | 0.03 |
|  3 | 997 | 0.11 | 0.03 |
|  4 | 9973 | 0.11 | 0.03 |
|  5 | 99991 | 0.11 | 0.03 |
|  6 | 999983 | 0.11 | 0.03 |
|  7 | 9999991 | 0.11 | 0.03 |
|  8 | 99999989 | 0.11 | 0.03 |
|  9 | 999999937 | 0.11 | 0.03 |
| 10 | 9999999967 | 0.11 | 0.04 |
| 11 | 99999999977 | 0.11 | 0.09 |
| 12 | 999999999989 | 0.12 | 0.35 |
| 13 | 9999999999971 | 0.14 | 2.25 |
| 14 | 99999999999973 | 0.22 | 19.04 |
| 15 | 999999999999989 | 0.48 | 185.16 |
| 16 | 9999999999999937 | 1.30 | 1835.94 |
| 17 | 99999999999999997 | 3.94 | 4869.46 |
| 18 | 999999999999999989 | 12.32 | ? |
| 19 | 9999999999999999961 | 43.02 | ? |

[factor-06.txt](https://github.com/user-attachments/files/16480719/factor-06.txt)

**Comment by pjacklam** [comment]
And here is the correct version of my version of the `factor` program: [factor-06.txt](https://github.com/user-attachments/files/16480750/factor-06.txt)

**Comment by briandfoy** [comment]
If you'd like to make a pull request to completely replace the existing one, go for it. There's a little bit of metadata in pod at the top of each existing program that you can add and you'd be all set.

If you don't want to do that, I'll take care of it. If nothing happens this week, just ping me to do it.

**Comment by pjacklam** [comment]
I have created a pull request. I have hardly ever done this before, so please be kind. :-) You are of course free to modify the code, comments, and the rest as you wish.

**Comment by briandfoy** [comment]
Merged in #700
